[WIP] Effects Refactoring

kshitij98 opened the pull request:
The pull request is at a very initial stage and a lot of work still needs to be done on it.

The current scope of this PR is:
- [x] Consolidate `EffectChain` and `EffectChainSlot` class
- [x] Remove EffectChainManager
- [x] Remove EngineEffectRack
- [x] Remove EffectRack
- [x] Consolidate Effect/EffectSlot
- [x] Remove unnecessary code from EffectParameter
- [x] Implement parameter hiding and rearrangement (skip implementing GUI for now)
- [ ] Reimplement saving/loading from XML with new EffectChainPreset & EffectPreset classes.


<br/>

There is some confusion with the current naming of classes. The newly proposed class names are:
- [ ] EffectState -> EffectProcessorState
- [ ] EffectChain/EffectChainSlot -> EffectRack
- [ ] EngineEffectChain -> EngineEffectRack (loaded using EffectRackPreset)
- [ ] Effect/EffectSlot -> EffectSlot
- [ ] EngineEffect -> EngineEffectSlot (loaded using EffectPreset)
- [ ] EffectProcessor -> EffectProcessor

<i><b>NOTE</b>: XML loading and saving has been commented out and will be re-implemented with the new `Preset` classes</i>

kshitij98 committed:
Consolidated EffectChain and EffectChainSlot

- EffectChain and EffectChainSlot had no clear separation of responsibility
- Had to keep the duplicated state in both the classes in sync and thus making it more susceptible to bugs

  NOTE: XML saving/loading of effects is broken and will be re-implemented in the new architecture

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -165,102 +583,175 @@ void EffectChainSlot::slotChainChannelStatusChanged(const QString& group,
 
 void EffectChainSlot::slotChainEffectChanged(unsigned int effectSlotNumber,
                                              bool shouldEmit) {
```
With the EffectChain/EffectChainSlot classes consolidated, let's work on removing these overcomplicated slotChain* functions.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -116,10 +144,400 @@ EffectChainSlot::~EffectChainSlot() {
+
+void EffectChainSlot::enableForInputChannel(const ChannelHandleAndGroup& handle_group) {
+    // TODO(Be): remove m_enabledChannels from this class and move this logic
+    // to EffectChainSlot
```
Let's take care of this TODO. The `updateRoutingSwitches` function should be removed. That function was an ugly hack that was only necessary because of the split between EffectChain and EffectChainSlot.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -165,102 +583,175 @@ void EffectChainSlot::slotChainChannelStatusChanged(const QString& group,
 
+// TODO(Kshitij) : Return EffectChainPreset instead of EffectChain
+// IMPORTANT(Kshitij) : Check what to return and keep the code working
 EffectChainPointer EffectChainSlot::getOrCreateEffectChain(
```
This function was another ugly piece of duct tape to keep the old architecture from crumbling. Let's remove it.

Be-ing reviewed (changes requested):
This looks like a decent start. Untangling the mess from the old architecture won't happen all at once, so I don't expect you to take care of everything I am pointing out now in the next commit.

In general, wherever the old EffectChainSlot class called a function of m_pEffectChain, remove the old function of the EffectChain class and move its contents to the function of the old EffectChainSlot class where it was called.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -824,6 +845,14 @@ void EffectChainSlot::registerInputChannel(const ChannelHandleAndGroup& handle_g
+        enableForInputChannel(pInfo->handle_group);
+    } else {
+        disableForInputChannel(pInfo->handle_group);
+    }
```
I think you could call slotChannelStatusChanged here instead of duplicating code.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -447,6 +449,11 @@ void EffectChainSlot::replaceEffect(unsigned int effectSlotNumber,
         if (m_bAddedToEngine) {
             qDebug() << "Adding " << pEffect->getManifest()->id() << " at " << effectSlotNumber;
+            qDebug() << "INPUT CHANNELS: ";
+            for (auto channel : m_enabledInputChannels) {
```
m_enabledInputChannels should be removed. It is redundant with the ChannelInfos' ControlObjects.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -114,12 +144,408 @@ EffectChainSlot::~EffectChainSlot() {
+        m_pEffectsManager->writeRequest(request);
+
+        // emit(channelStatusChanged(handle_group.name(), false));
+        slotChainChannelStatusChanged(handle_group.name(), false);
```
I think this code follows a circular path now. All that `slotChainChannelStatusChanged` does is update EffectChainSlot's ChannelInfo, but `enableForInputChannel` and `disableForInputChannel` should only be called by `slotChannelStatusChanged` when the ChannelInfo's ControlObject emits its `valueChanged` signal.

Be-ing commented on src/effects/effectchainslot.h:
```diff
@@ -53,6 +70,79 @@ class EffectChainSlot : public QObject {
+
+
+    // START EffectChain
+    void updateEngineState();
```
This should be private.

Be-ing commented on src/effects/effectchainslot.h:
```diff
@@ -53,6 +70,79 @@ class EffectChainSlot : public QObject {
+    const QSet<ChannelHandleAndGroup>& enabledChannels() const;
+    void disableForInputChannel(const ChannelHandleAndGroup& handle_group);
+
+    EffectChainPointer prototype() const;
```
This function doesn't make sense with the new architecture. Remove it.

Be-ing commented on src/effects/effectrack.cpp:
```diff
@@ -105,8 +127,11 @@ EffectChainSlotPointer EffectRack::getEffectChainSlot(int i) {
     return m_effectChainSlots[i];
 }
 
+// NOTE(Kshitij) : Why are we using a clone of the EffectChain?
```
This concept of cloning chains does not make sense with the new architecture. Comment out this function for now and add a TODO comment noting to come back to it when we implement loading chain presets from the user's list of saved presets. Likewise for EffectRack::loadPrevChain.

Be-ing replied:
Actually these functions don't belong in EffectRack. They should be functions of EffectChainSlot. There's no need for the EffectRack layer and it should be removed.

Be-ing commented on src/effects/effectrack.cpp:
```diff
@@ -157,9 +184,15 @@ void EffectRack::maybeLoadEffect(const unsigned int iChainSlotNumber,
+        // NOTE(Kshitij) : Replaced with new effect
+        // EffectChainPointer pChain = pChainSlot->getOrCreateEffectChain(m_pEffectsManager);
+        // EffectPointer pEffect = m_pEffectsManager->instantiateEffect(id);
+        // pChain->replaceEffect(iEffectSlotNumber, pEffect);
```
remove commented out code

Be-ing commented on src/effects/effectrack.cpp:
```diff
@@ -174,9 +207,14 @@ void EffectRack::loadNextEffect(const unsigned int iChainSlotNumber,
+    // NOTE(Kshitij) : Replaced with new effect
+    // EffectChainSlotPointer pChainSlot = m_effectChainSlots[iChainSlotNumber];
+    // EffectChainPointer pChain = pChainSlot->getOrCreateEffectChain(m_pEffectsManager);
+    // pChain->replaceEffect(iEffectSlotNumber, pNextEffect);
```
remove commented out code

Be-ing commented on src/effects/effectrack.cpp:
```diff
@@ -214,8 +257,11 @@ QDomElement EffectRack::toXml(QDomDocument* doc) const {
+        // NOTE(Kshitij) : refreshing all chainslots than effect chains
+        // EffectChainPointer pChain = pChainSlot->getOrCreateEffectChain(m_pEffectsManager);
+        // pChain->refreshAllEffects();
+        // EffectChainPointer pChain = pChainSlot->getOrCreateEffectChain(m_pEffectsManager);
```
remove commented out code

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -114,12 +144,408 @@ EffectChainSlot::~EffectChainSlot() {
+    pRequest->pTargetChain = m_pEngineEffectChain;
+    pRequest->SetEffectChainParameters.enabled = m_bEnabled;
+    pRequest->SetEffectChainParameters.mix_mode = m_mixMode;
+    pRequest->SetEffectChainParameters.mix = m_dMix;
```
Storing these states in private member variables is redundant with the ControlObjects. We can get rid of this redundant variables and all their setter/getter functions by making this function a slot and connecting it to the `valueChanged` signal of the ControlObjects.

Be-ing replied:
I think this should be your next step of cleanup.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -362,28 +887,36 @@ void EffectChainSlot::slotControlClear(double v) {
         qWarning() << debugString() << "value out of limits";
         v = math_clamp(v, 0.0, 1.0);
         m_pControlChainMix->set(v);
     }
```
ControlPotmeter makes sure the value stays in bounds already, so this code is redundant.

Be-ing commented on src/effects/effectsmanager.cpp:
```diff
@@ -347,63 +349,65 @@ void EffectsManager::setup() {
-    // populate rack and restore state from effects.xml
-    m_pEffectChainManager->loadEffectChains();
-}
+    // NOTE(Kshitij) : Why are we adding single-effect effect chains on startup?
```
This was a really ugly hack around some of the skins in the initial implementation of the effects system in Mixxx 2.0 not showing chains at all. Remove this outdated code please.

Be-ing commented:
Please make your next commit simply removing commented out and outdated code. Then it will be easier to review the code you actually change in following commits.

Be-ing commented on src/effects/effectrack.cpp:
```diff
@@ -309,8 +267,9 @@ OutputEffectRack::OutputEffectRack(EffectsManager* pEffectsManager,
-    pChain->setMix(1.0);
+
+    pChainSlot->enableForInputChannel(*masterHandleAndGroup);
+    pChainSlot->setMix(1.0);
```
There should not be any public methods of EffectChainSlot for setting enabled input channels or the mix knob. Add a TODO comment noting this and explaining we can resolve this by making these subclasses of EffectChainSlot and removing the EffectRack layer.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -113,13 +136,305 @@ EffectChainSlot::~EffectChainSlot() {
+    // not needed.
+    if (bWasAlreadyEnabled) {
+        return;
+    }
```
Can this check be removed at this point? IIRC it was only necessary before because of the convoluted initialization process.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -429,15 +685,14 @@ void EffectChainSlot::slotControlChainPrevPreset(double v) {
+    if (pChannelInfo != NULL && pChannelInfo->pEnabled != NULL) {
+        bool bEnable = pChannelInfo->pEnabled->toBool();
+        if (bEnable) {
+            enableForInputChannel(pChannelInfo->handle_group);
```
Let's get rid of these functions and move their code into this one.

Be-ing replied:
Ping. Either move these into one big function here or move the code for `enableForInputChannel` and `disableForInputChannel` next to this function in the file.

Be-ing commented:
I am repeating the plan we came up with in our private Zulip chat:
1. Finish simplifying EffectChainSlot
2. Remove EffectChainManager
3. Remove EffectRack/EngineEffectRack
4. Consolidate Effect/EffectSlot
5. Reimplement saving/loading from XML with new EffectChainPreset & EffectPreset classes.

daschuer commented:
I think we will need the EffectRack layer.

Be-ing commented:
@kshitij98 did you change the email you configured to be associated with your Git commits? The new commits are showing as being from a different GitHub user.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -139,18 +422,9 @@ void EffectChainSlot::slotChainNameChanged(const QString&) {
-
-void EffectChainSlot::slotChainMixModeChanged(EffectChainMixMode mixMode) {
-    m_pControlChainMixMode->set(static_cast<double>(mixMode));
+void EffectChainSlot::slotChainUpdated(double v) {
```
I think this function is unnecessary. Can't you just connect the signals to `sendParameterUpdate`?

kshitij98 commented:
> @kshitij98 did you change the email you configured to be associated with your Git commits? The new commits are showing as being from a different GitHub user.

My laptop was not working fine yesterday so I had to use a friend's pc for the same. I forgot to change the config setting of git. Should I reset the author in the previous commits?

Be-ing commented:
When you're done consolidating EffectChain/EffectChainSlot, rebase and squash your work down to one commit to remove the temporary intermediate code from the Git history. At that time you can remove the other email associated with the new commits. For now let's just keep going.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -206,18 +207,13 @@ void EffectChainSlot::setDescription(const QString& description) {
-    }
+    // TODO(Kshitij) : remove
+    qDebug() << "Enabling " << handle_group.name();
+    clock_t start = clock();
```
Use the ScopedTimer class from `util/timer.h`

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -258,23 +254,30 @@ void EffectChainSlot::enableForInputChannel(const ChannelHandleAndGroup& handle_
+    qDebug() << "ENABLED " << handle_group.name();
 }
 
 bool EffectChainSlot::enabledForChannel(const ChannelHandleAndGroup& handle_group) const {
```
This function is unused and this information shouldn't be exposed to other classes. Remove it.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -206,16 +206,7 @@ void EffectChainSlot::setDescription(const QString& description) {
-    // The allocation of EffectStates below may be expensive, so avoid it if
-    // not needed.
-    if (bWasAlreadyEnabled) {
+    if (m_enabledInputChannels.contains(handle_group)) {
```
change this to a VERIFY_OR_DEBUG_ASSERT because it should never happen

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -258,23 +249,20 @@ void EffectChainSlot::enableForInputChannel(const ChannelHandleAndGroup& handle_
-        m_pEffectsManager->writeRequest(request);
-
-        slotChainChannelStatusChanged(handle_group.name(), false);
+    if (!m_enabledInputChannels.remove(handle_group)) {
```
likewise change this to a VERIFY_OR_DEBUG_ASSERT

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -113,184 +133,321 @@ EffectChainSlot::~EffectChainSlot() {
+    if (!m_enabledInputChannels.remove(handle_group)) {
+        return;
     }
+    
```
There are extra spaces on this line. Please configure your editor to automatically remove trailing spaces when you save a file. In KDevelop you can do this in Settings > KDevelop > Editor > Open/Save > Automatic Cleanups on Save > Remove trailing spaces on modified lines.

kshitij98 commented on src/effects/effectchainslot.cpp:
```diff
@@ -116,10 +144,400 @@ EffectChainSlot::~EffectChainSlot() {
+    m_pEngineEffectChain = nullptr;
+}
+
+void EffectChainSlot::updateEngineState() {
```
I can make this function private only when any other non-inheriting class is not class is not calling `updateEngineState`

I have added a TODO to make this private after the EffectRack layer is removed.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -113,190 +136,315 @@ EffectChainSlot::~EffectChainSlot() {
-        EffectSlotPointer pSlot;
-        EffectPointer pEffect;
+void EffectChainSlot::setMix(const double& dMix) {
+    m_pControlChainMix->set(dMix);
```
The `valueChanged` signal is not emitted when set from the class the ControlObject is in (I do not quite understand why. @daschuer can you explain why this happens?). Adding `sendParameterUpdate()` here fixes the PerGroupRacks and OutputEffectRack.

Be-ing commented:
The regular effects are not being applied if they are enabled for an input channel on startup. After toggling the routing switch they work. I am trying to debug this...

Be-ing commented:
I have [sent you a pull request](https://github.com/kshitij98/mixxx/pull/1) to fix the initialization process.

Be-ing commented on src/effects/effectchainslot.h:
```diff
@@ -5,36 +5,41 @@
 
+class EffectsManager;
+class EngineEffectRack;
+class EngineEffectChain;
```
These forward declarations aren't needed. Remove them.

Be-ing commented on src/effects/effectchainslot.h:
```diff
@@ -53,30 +58,77 @@ class EffectChainSlot : public QObject {
+    QString description() const;
+    void setDescription(const QString& description);
+
+    // TODO(Kshitij) : Make this function private once EffectRack layer is removed
```
This function should be removed when the EffectRack layer is removed, not made private.

Be-ing commented on src/effects/effectchainslot.h:
```diff
@@ -53,30 +58,77 @@ class EffectChainSlot : public QObject {
+    void disableForInputChannel(const ChannelHandleAndGroup& handle_group);
+
+    // TODO(Kshitij) : Make this function private once EffectRack layer is removed
+    void updateEngineState();
```
This function isn't needed. Remove it.

Be-ing commented:
EffectChainSlot is down to about 600 lines of code :D

Be-ing commented:
I am thinking we should rename EffectChainSlot to EffectChain now. What do you think?

Be-ing commented:
EffectChainSlotTest doesn't make sense with the new architecture and IMO was never really useful anyway. Delete it.

Be-ing commented:
> I am thinking we should rename EffectChainSlot to EffectChain now. What do you think?

I am now leaning towards renaming it to EffectUnit. "EffectChain" is somewhat misleading because multiple units can be chained together by assigning them to the same input channel. ([Related Zulip discussion](https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/chaining.20.3E3.20effects.20in.20D.2BW.20mode))

kshitij98 commented:
> > I am thinking we should rename EffectChainSlot to EffectChain now. What do you think?

> I am now leaning towards renaming it to EffectUnit. "EffectChain" is somewhat misleading because multiple units can be chained together by assigning them to the same input channel. (Related Zulip discussion)

Yes, makes sense. We should rename it to something like `EffectUnit` or `EffectModule`. I vote for `EffectUnit`.

Be-ing commented:
I don't think "effect module" is a term used in other audio contexts, so let's go with "effect unit", although it is somewhat of a misnomer because a "unit" typically means one discrete thing, but this is a group of 3 things.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -113,190 +136,199 @@ EffectChainSlot::~EffectChainSlot() {
+    return m_effects;
+}
+
+EngineEffectChain* EffectChainSlot::getEngineEffectChain() {
```
This function is unused. Delete it.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -113,190 +136,199 @@ EffectChainSlot::~EffectChainSlot() {
+
+    m_pControlNumEffects->forceSet(math_min(
+            static_cast<unsigned int>(m_slots.size()),
+            numEffects()));
```
This is the only place the `numEffects` function is used. Remove that function and replace it with simply `m_effects.size()`.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -113,190 +136,199 @@ EffectChainSlot::~EffectChainSlot() {
 
 EffectSlotPointer EffectChainSlot::addEffectSlot(const QString& group) {
-    //qDebug() << debugString() << "addEffectSlot" << group;
+    qDebug() << debugString() << "addEffectSlot" << group;
```
comment this out again

Be-ing commented on src/effects/effectrack.cpp:
```diff
@@ -332,20 +274,19 @@ void PerGroupRack::setupForGroup(const QString& groupName) {
     int iChainSlotNumber = m_groupToChainSlot.size();
     QString chainSlotGroup = formatEffectChainSlotGroupForGroup(
         getRackNumber(), iChainSlotNumber, groupName);
+    qDebug() << "Chain slot group = " << chainSlotGroup;
```
comment this out

Be-ing commented on src/widget/weffectchain.cpp:
```diff
@@ -35,14 +35,12 @@ void WEffectChain::setEffectChainSlot(EffectChainSlotPointer pEffectChainSlot) {
 }
 
 void WEffectChain::chainUpdated() {
+    qDebug() << "chainUpdated()";
```
comment out

kshitij98 commented:
How about EffectChainUnit, other than being a little verbose, it seems to create no confusion.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -1,22 +1,40 @@
+          m_name(""),
+          m_description(""),
+          m_pEngineEffectChain(nullptr) {
+    qDebug() << "EffectChainSlot::EffectChainSlot " << pRack << ' ' << group << ' ' << iChainNumber;
```
comment out

Be-ing commented:
Down to ~500 lines of code if you don't count the commented out XML functions :)

daschuer commented:
> Yes, makes sense. We should rename it to something like EffectUnit or EffectModule. I vote for EffectUnit.

We should not cycle the names around. It will be a great source of confusion because old references. 
No one will be sure if the old EffectUnit is meant or the old. In this case, it is better to stick to legacy.

Be-ing commented:
The term "effect chain" has never been used in Mixxx's GUI. What has been called an "effect chain" in the code has been called an "effect unit" in the ControlObject names, which creates confusion.

Be-ing commented:
The [manual](https://mixxx.org/manual/2.1/en/chapters/effects.html) also calls them "effect units".

daschuer commented:
The effect unit was the facility to hold an effect chain. The old nomenclature was very consistent, but probably hard to understand. I am a bit afraid this PR does too much.

Be-ing commented:
That is consistent with renaming what is currently called EffectChainSlot in this branch to EffectUnit.

kshitij98 commented:
I think we should refer to https://github.com/mixxxdj/mixxx/blob/master/src/skin/tooltips.cpp

Both the terms 'unit' and 'chain' are being used here and I think we must settle on one.

Be-ing commented on src/test/metaknob_link_test.cpp:
```diff
@@ -1,263 +0,0 @@
-#include "util/time.h"
-#include "control/controlproxy.h"
-
-class MetaLinkTest : public BaseEffectTest {
```
These tests are still helpful. Bring them back.

Be-ing commented on src/test/effectslottest.cpp:
```diff
@@ -1,80 +0,0 @@
-using ::testing::Return;
-using ::testing::_;
-
-class EffectSlotTest : public BaseEffectTest {
```
This is unrelated to the work you've done so far, so leave these tests in for now. When we consolidate Effect & EffectSlot, then let's remove these tests.

kshitij98 committed:
Removed EffectChainManager layer

- Removed the superfluous class `EffectChainManager` whose functionality is now shifted to `EffectsManager`
- Updated EffectProcessor to fix the circular include problem during compilation

  NOTE: XML saving/loading of effects is broken and will be re-implemented in the new architecture

Be-ing commented:
Please rebase this on master to squash everything before 145befd29663c3e638414b94a969486052a78af3 to one commit.

Be-ing commented:
I sent you a [pull request](https://github.com/kshitij98/mixxx/pull/2) to fix the build issues from removing EffectChainManager.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -429,17 +420,75 @@ void EffectChainSlot::slotControlChainPrevPreset(double v) {
+          mixxx::AudioSignal::SampleRate(96000),
+          MAX_BUFFER_LEN / mixxx::kEngineChannelCount);
+
+    for (int i = 0; i < m_effects.size(); ++i) {
```
This should be as simple as looping over all the EffectSlots and calling a method for them to create an EffectState for the input channel. I started working on this, but unfortunately it is a bit of a pain right now and will require more work later because of the separation between EffectSlot and Effect. For now let's leave it as a TODO.

kshitij98 commented:
> I sent you a pull request to fix the build issues from removing EffectChainManager.

Thank you! I have rebased your commits into this branch (I had to cherry-pick your commits because I had already sqaushed the earlier commits into one) and squashed them into a single [commit](https://github.com/mixxxdj/mixxx/pull/1705/commits/104a2fb3b91af6fcc4f333164fa4900ba18cac49)

Be-ing commented:
Removing EffectChainManager made effectmanager.cpp about 100 lines longer. IMO this is worth a simpler architecture. :+1:

Be-ing commented:
The Git history looks good now, thank you! :+1:

Be-ing commented:
Oh wait, please amend both commit messages so far to mention that XML saving/loading is temporarily not working and will be reimplemented after more refactoring.

kshitij98 committed:
Removed EngineEffectRack layer

- Removed the superfluous layer `EngineEffectRack` above all the effect chains in the engine

  NOTE: XML saving/loading of effects is broken and will be re-implemented in the new architecture

daschuer commented:
Unfortunately we have no consensus about removing the EngineEffectRack layer.  I think it is better to not do this in this PR to risk not to merge it. I think it is finally only a renaming right?

Be-ing commented:
I've already explained many times why the EffectRack layer is being removed. I will not get into a prolonged discussion about this on this PR. We do not have time in GSOC to delay a student's project for a long discussion about an implementation detail irrelevant to users.

daschuer commented:
I think we had a good discussion on Zulip. I have fully understand your view, but we have not found a decision yet.   

The remaining point is: Should we change the CO namespace, which is a very user visible change.
I think we have agreed that the CO name-space names should be reflected the internal class names. 
So we cannot remove the EngineEffectRack layer without a decision about the CO namespace. 

I like to avoid that this decision blocks this PR and the student project. Thats why the EffectRack layer remove should not be part of this, to avoid the blocking. We can do it at any later time. 
I think it is finally just a renaming of classes and name-space.

Be-ing commented:
We can remove the superfluous classes from the code regardless of whether we rename the ControlObjects. We can continue this discussion on Zulip, but please, let's not clutter this PR with the discussion.

Be-ing commented:
Please stop cluttering this PR with this. PRs are not good places for extended discussion, Zulip is.

daschuer commented:
PRs are for discussing commits and code changes. I think we should revert the latest commit until we have found a conclusion on Zulip, otherwise we cannot merge this. On Zulip this info will be out of sight.

Be-ing commented:
It is not fair to @kshitij98 to hold up his project for this or continue adding noise to this PR.

kshitij98 committed:
Removed EffectRack layer

- Removed the superfluous layer `EffectRack` above all the effect chain slots
- The subclasses of former `EffectRack` class are now subclasses of `EffectChainSlot`

  NOTE: XML saving/loading of effects is broken and will be re-implemented in the new architecture

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -182,6 +182,10 @@ void EffectChainSlot::removeFromEngine(int iIndex) {
     m_pEngineEffectChain = nullptr;
 }
 
+QString EffectChainSlot::formatEffectChainSlotGroup(const QString& group) const {
```
This should take an integer.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -275,6 +274,10 @@ QString EffectChainSlot::id() const {
     return m_id;
 }
 
+QString EffectChainSlot::group() const {
```
Is this function necessary?

kshitij98 replied:
I am using this function to fetch the formatted group name of the effect chain slot. Otherwise, I am constructing the Effect Chain Slot by passing only the chain slot number or group.

Be-ing replied:
okay

Be-ing commented on src/effects/specialeffectchainslots.cpp:
```diff
@@ -0,0 +1,45 @@
+		  				  iChainNumber, pEffectsManager,
+		  				  SignalProcessingStage::Prefader, false,
+		  				  QString("[QuickEffectRack1_%1]").arg(group)) {
+	// setMix(1.0);
```
```
m_pControlChainMix->set(1.0);
sendParameterUpdate();
```

Be-ing commented on src/effects/specialeffectchainslots.cpp:
```diff
@@ -0,0 +1,45 @@
+		  				  SignalProcessingStage::Prefader, false,
+		  				  QString("[QuickEffectRack1_%1]").arg(group)) {
+	// setMix(1.0);
+	// registerInputChannel();
```
I think you need to pass a ChannelHandleAndGroup to these EffectChainSlot subclass constructors from EffectsManager. You'll have to call `enableInputChannel` too.

Be-ing reviewed (changes requested):
Looks like good progress. Next steps will be instantiating the new EffectChainSlot subclasses in EffectsManager then deleting EffectRack.

Be-ing commented on src/effects/specialeffectchainslots.cpp:
```diff
@@ -2,44 +2,180 @@
-	return QString("[OutputEffectRack_%1]").arg(group);
+void EqualizerEffectChainSlot::configureForGroup(const QString& group) const {
+    // Create aliases for legacy EQ controls.
+    // NOTE(rryan): If we ever add a second EqualizerRack then we need to make
```
remove this outdated comment

Be-ing commented:
There are a lot of member variables and methods of EffectChainSlot that are private but now I realize they should be protected.

kshitij98 commented:
Yes, will work on that in the next commit.

Be-ing commented on src/effects/specialeffectchainslots.cpp:
```diff
@@ -0,0 +1,201 @@
+    DEBUG_ASSERT(handleAndGroup != nullptr);
+
+    // Register this channel alone with the chain slot.
+    registerInputChannel(*handleAndGroup);
```
I think `enableForInputChannel` is needed here.

Be-ing commented on src/effects/specialeffectchainslots.cpp:
```diff
@@ -0,0 +1,201 @@
+    }
+}
+
+QString StandardEffectChainSlot::formatEffectChainSlotGroup(const QString& group) {
```
This function is redundant with `formatEffectChainSlotGroupString`

daschuer commented:
Can we cut out a branch, that does not break storing the effect states?
Which is the first commit that breakers it?
I think that will be convenient to merge and review.

Be-ing commented:
XML saving/loading was broken in the first commit. The purpose of this refactoring is to simplify XML saving/loading and decouple it from the initialization process. Reimplementing it before the refactoring is done would be a waste of time.

Be-ing commented on src/effects/specialeffectchainslots.cpp:
```diff
@@ -0,0 +1,197 @@
+}
+
+QString EqualizerEffectChainSlot::formatEffectChainSlotGroup(const QString& group) {
+    return formatEffectChainSlotGroupString(group);
```
This is overcomplicated. Can't you merge `formatEffectChainSlotGroupString` and `formatEffectChainSlotGroup` into one function?

kshitij98 replied:
The issue I'm facing is that we cannot have static virtual functions. So currently I've kept both. One solution is we can remove the virtual functions and not call any of the formatting functions in the base class.

Be-ing replied:
Why do we need static functions?

Be-ing replied:
Oh I see, some of the static functions are used by ControlPickerMenu... :/ Hmm... there must be a better way to organize this...

Be-ing replied:
I think your idea to not call any of the formatting functions in the base PerGroupEffectChainSlot class would be the best way around this.

kshitij98 replied:
There are calls to the formatting functions in ControlPickerMenu and DlgPrefEQ without instantiating an EffectChainSlot. The other workaround can be to make temporary EffectChainSlots to use the formatting functions.

Be-ing commented on src/effects/specialeffectchainslots.cpp:
```diff
@@ -0,0 +1,197 @@
+    enableForInputChannel(*handleAndGroup);
+
+    // Add a single effect slot
+    addEffectSlot(chainSlotGroup);
```
This is not working. The Effect group name is not the same as the ChainSlot group name. Use `formatEffectSlotGroup`

Be-ing commented on src/effects/specialeffectchainslots.cpp:
```diff
@@ -0,0 +1,197 @@
+    setSuperParameterDefaultValue(0.5);
+}
+
+QString QuickEffectChainSlot::formatEffectSlotGroup(const QString& group) {
```
This assumes that there will only ever be one effect in each QuickEffect chain. After we implement saving/loading custom chains, I would like it to be possible to load a custom chain preset for the QuickEffects, so that assumption would get in the way.

Be-ing commented on src/effects/specialeffectchainslots.cpp:
```diff
@@ -0,0 +1,197 @@
+                .arg(group);
+}
+
+QString QuickEffectChainSlot::formatEffectSlotGroupString(const QString& group) {
```
This function is redundant.

Be-ing commented on src/effects/specialeffectchainslots.cpp:
```diff
@@ -71,6 +77,10 @@ PerGroupEffectChainSlot::PerGroupEffectChainSlot(const QString& group,
+    setup(group);
+}
+
+void PerGroupEffectChainSlot::setup(const QString& group) {
```
Why did you split this out from the constructor?

kshitij98 replied:
We cannot call virtual functions inside the constructor/destructor. It returns an error if we try to do the same. This is the same reason I'm passing the chainSlotGroup and the group both to the base class PerGroupEffectChainSlot.

Reference: https://stackoverflow.com/questions/962132/calling-virtual-functions-inside-constructors

Be-ing replied:
I think this won't be a problem if we move the call to `addEffectSlot` into the EqualizerEffectChainSlot and QuickEffectChainSlot constructors instead of the base PerGroupEffectChainSlot constructor. That will also allow adding more effect slots for the QuickEffects than the EQs.

kshitij98 commented on src/effects/effectsmanager.cpp:
```diff
@@ -212,40 +232,113 @@ EffectPointer EffectsManager::instantiateEffect(const QString& effectId) {
+    int iChainSlotNumber = m_equalizerEffectChainSlots.size();
+
+    auto pChainSlot = EqualizerEffectChainSlotPointer(
+            new EqualizerEffectChainSlot(groupName, iChainSlotNumber, this));
```
I think we still would have to maintain the `m_iChainNumber` for passing it to the EffectSlot which would use it to format the group string. I think we can only remove `AddEffectChain.iIndex` from the request.

Be-ing replied:
Only StandardEffectChainSlot needs it for formatting effect group strings. The other EffectChainSlot subclasses do not. So I think `m_iChainNumber` should be removed from the base EffectChainSlot class.

Be-ing commented on src/test/effectslottest.cpp:
```diff
@@ -31,19 +30,18 @@ class EffectSlotTest : public BaseEffectTest {
-    // StandardEffectRack::addEffectChainSlot automatically adds 4 effect
+    m_pEffectsManager->addStandardEffectChainSlots();
+    EffectChainSlotPointer pChainSlot = m_pEffectsManager->getStandardEffectChainSlot(iChainNumber);
+    // StandardEffectChainSlot::addEffectChainSlot automatically adds 4 effect
```
Remove this out-of-date comment.

Be-ing commented on src/effects/effectchainslot.h:
```diff
@@ -35,7 +34,8 @@ class EffectChainSlot : public QObject {
 
-    void registerInputChannel(const ChannelHandleAndGroup& handle_group);
+    void registerInputChannel(const ChannelHandleAndGroup& handle_group,
+                              const double initialValue = 0.0);
```
Adding a new parameter to this function with a default value is a clever way to implement this. :+1: I was thinking that logic should be moved to StandardEffectRack anyway.

Be-ing reviewed (approved):
Great, I think we're ready to squash the EffectRack removal down to one commit. Leave 4987feff9bbab5490e22d5ccda6ebda588bf28e9 as a separate commit though.

Be-ing commented:
Between 4987feff9bbab5490e22d5ccda6ebda588bf28e9 and f45d0005c736d50cfb12c3773604410765b52744, a net of 803 lines of code were removed! :tada:

kshitij98 committed:
Consolidated Effect and EffectSlot

- Effect and EffectSlot classes had no clear separation of responsibility
- Had to keep the duplicated state in both the classes in sync and thus was more susceptible to bugs
- Ensured a loaded effect slot to have a corresponding EngineEffect added to the engine

  NOTE: XML saving/loading of effects is broken and will be re-implemented in the new architecture

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -148,71 +309,83 @@ EffectButtonParameterSlotPointer EffectSlot::getEffectButtonParameterSlot(unsign
-        m_pControlLoaded->forceSet(1.0);
-        m_pControlNumParameters->forceSet(pEffect->numKnobParameters());
-        m_pControlNumButtonParameters->forceSet(pEffect->numButtonParameters());
+void EffectSlot::loadEffectToSlot(EffectsManager* pEffectsManager,
```
This class already has a pointer to EffectsManager

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -89,10 +100,84 @@ EffectSlot::~EffectSlot() {
+
+    m_pEngineEffect = new EngineEffect(m_pManifest,
+            activeInputChannels,
+            pEffectsManager,
```
m_pEffectsManager

Be-ing commented on src/effects/effectsbackend.cpp:
```diff
@@ -45,15 +45,17 @@ bool EffectsBackend::canInstantiateEffect(const QString& effectId) const {
-EffectPointer EffectsBackend::instantiateEffect(EffectsManager* pEffectsManager,
-                                                const QString& effectId) {
+bool EffectsBackend::instantiateEffect(EffectsManager* pEffectsManager,
+        const QString& effectId, const EffectSlotPointer pEffectSlot,
```
I'm confused why this function takes an EffectSlotPointer now. Shouldn't it return an EngineEffect which then gets loaded into the EffectSlot?

Be-ing reviewed (changes requested):
This looks like a pretty good start. I'm confused about the code path for loading effects though.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -192,37 +192,21 @@ void EffectChainSlot::setDescription(const QString& description) {
-        loadNew = true;
-    } else if (id != pEffectSlot->getManifest()->id()) {
-        loadNew = true;
+    if (pEffectSlot == nullptr) {
```
Use VERIFY_OR_DEBUG_ASSERT macro

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -192,37 +192,21 @@ void EffectChainSlot::setDescription(const QString& description) {
-        VERIFY_OR_DEBUG_ASSERT(m_pEffectsManager->instantiateEffect(id, pEffectSlot, m_enabledInputChannels)) {
-            qWarning() << "Unable to load effect.";
+    if (pEffectSlot->getManifest() == nullptr || pEffectSlot->getManifest()->id() != id) {
+        VERIFY_OR_DEBUG_ASSERT(pEffectSlot->loadEffect(pManifest, pInstantiator, m_enabledInputChannels)) {
```
Remove this VERIFY_OR_DEBUG_ASSERT and make EffectSlot::loadEffect return void. Any errors that happen when EffectSlot::loadEffect is called should be handled by VERIFY_OR_DEBUG_ASSERTs in that function.

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -309,39 +298,29 @@ EffectButtonParameterSlotPointer EffectSlot::getEffectButtonParameterSlot(unsign
-    // Removing the loaded effect
-    if (m_pManifest == nullptr) {
-        return;
+    VERIFY_OR_DEBUG_ASSERT(m_pManifest != nullptr) {
```
Should an effect be able to be unloaded by calling loadEffect with a null EffectManifestPointer and/or EffectInstantiatorPointer?

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -309,39 +298,29 @@ EffectButtonParameterSlotPointer EffectSlot::getEffectButtonParameterSlot(unsign
-                this, pEffectsManager, m_parameters.size(), pManifestParameter);
+                this, m_pEffectsManager, m_parameters.size(), pManifestParameter);
         m_parameters.append(pParameter);
         if (m_parametersById.contains(pParameter->id())) {
```
VERIFY_OR_DEBUG_ASSERT

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -386,6 +365,15 @@ void EffectSlot::clear() {
+        delete pParameter;
+    }
+
+    removeFromEngine();
```
Should EffectSlot::removeFromEngine ever be called from anywhere else? If not, maybe that function should be removed and its code moved into this function.

kshitij98 replied:
No, It is not being called from anywhere else. But I think we should keep it as it is to maintain the uniformity of having a add/remove to/from engine function in each effect layer.

Be-ing commented on src/effects/effectsmanager.cpp:
```diff
@@ -133,6 +133,55 @@ void EffectsManager::registerOutputChannel(const ChannelHandleAndGroup& handle_g
+void EffectsManager::loadEffect(EffectChainSlotPointer pChainSlot,
+        const int iEffectSlotNumber, const QString& effectId,
+        EffectBackendType backendType) {
+    foreach (EffectsBackend* pBackend, m_effectsBackends) {
```
Use standard C++ `for (EffectsBackend* pBackend : m_effectsBackends)` instead of Qt `foreach` macro... although this loop wouldn't be necessary if EffectsManager stored a QHash of EffectsBackends indexed by EffectBackendType.

Be-ing commented on src/preferences/dialog/dlgprefeq.cpp:
```diff
@@ -463,12 +463,13 @@ void DlgPrefEQ::applySelections() {
             VERIFY_OR_DEBUG_ASSERT(pChainSlot != nullptr);
 
-            pChainSlot->loadEffect(0, effectId);
+            m_pEffectsManager->loadEqualizerEffect(chainSlotGroup, 0, effectId, EffectBackendType::Unknown);
```
somehow we need to get the correct EffectBackendType here

Be-ing commented on src/effects/effectchainslot.h:
```diff
@@ -44,73 +48,82 @@ class EffectChainSlot : public QObject {
+    void clearChain(unsigned int iChainNumber, EffectChainSlotPointer pEffectChain);
 
     // Signal that indicates that the EffectChainSlot has been updated.
     void updated();
```
This signal name is quite unhelpful. Change it to `effectLoaded`. Be sure to update WEffectSelector.

kshitij98 replied:
I've set it to `effectChanged` because we are supposed to trigger this signal on unloading as well.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -429,81 +380,133 @@ void EffectChainSlot::slotControlChainPrevPreset(double v) {
+    qDebug() << "Effect Slots in EffectChainSlot" << m_effectSlots.size() << m_group;
+    for (int i = 0; i < m_effectSlots.size(); ++i) {
+        auto& statesMap = (*pEffectStatesMapArray)[i];
+        if (m_effectSlots[i] != nullptr) {
```
Change this to `if (m_effectSlots[i].isLoaded()) {`

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -91,10 +103,84 @@ EffectSlot::~EffectSlot() {
+}
+
+EffectState* EffectSlot::createState(const mixxx::EngineParameters& bufferParameters) {
+    if (m_pEngineEffect == nullptr) {
```
VERIFY_OR_DEBUG_ASSERT

Be-ing commented on src/effects/effectparameter.cpp:
```diff
@@ -2,14 +2,12 @@
+                                 int iParameterNumber, EffectManifestParameterPointer pParameter)
         : QObject(), // no parent
-          m_pEffect(pEffect),
+          m_pEffectSlot(pEffectSlot),
```
Change this to an `EngineEffect*`, then you can remove EffectSlot::getEngineEffect

Be-ing commented on src/effects/effectsmanager.cpp:
```diff
@@ -399,15 +399,24 @@ void EffectsManager::setup() {
 //     m_pEffectChainManager->loadEffectChains();
 // }
 
+inline void EffectsManager::refreshChainSlot(EffectChainSlotPointer pChainSlot) {
```
Why did you move this from EffectChainSlot?

kshitij98 replied:
Earlier the flow was DlgPrefLV2 -> EffectsManager::refreshAllChainSlots -> EffectChainSlot::refreshAllEffects -> EffectsManager::loadEffect -> EffectChainSlot::loadEffect -> EffectSlot::loadEffect. And the EffectsManager::loadEffect function required an Shared pointer of the EffectChainSlot. Moving this functionality to EffectsManager resolves the above issues and simplifies the flow.

Be-ing replied:
Okay, thank you for explaining. I see what the problem was, but I think we can come up with a better solution than this. EffectChainSlot::refreshAllEffects should not have needed to make a call to EffectsManager which required a pointer back to the EffectChainSlot which made the call. I think we can make an EffectSlot::refresh function if EffectSlot::loadEffect stores the EffectInstantiatorPointer it is passed as a member variable.

Be-ing commented on src/effects/effectchainslot.h:
```diff
@@ -34,6 +34,9 @@ class EffectChainSlot : public QObject {
 
     EffectSlotPointer addEffectSlot(const QString& group);
     EffectSlotPointer getEffectSlot(unsigned int slotNumber);
+    const unsigned int numEffectSlots() const {
```
This information shouldn't need to be exposed.

kshitij98 replied:
I have used this in EffectsManager to refresh all the effect slots. We can remove this if we change the refreshing methods.

Be-ing commented on src/effects/effectchainslot.h:
```diff
@@ -34,6 +34,9 @@ class EffectChainSlot : public QObject {
 
     EffectSlotPointer addEffectSlot(const QString& group);
```
make protected

Be-ing replied:
You can make this a separate commit then use `git rebase -i` to squash it into the EffectRack removal commit by marking it as "fixup" commit.

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -150,72 +312,85 @@ EffectButtonParameterSlotPointer EffectSlot::getEffectButtonParameterSlot(unsign
-        // EffectPointer to the EffectSlot. Effects and EngineEffects default to
-        // disabled, so if this EffectSlot was enabled, enable the Effect and EngineEffect.
-        pEffect->setEnabled(m_pControlEnabled->toBool());
+    VERIFY_OR_DEBUG_ASSERT(pManifest != nullptr && pInstantiator != nullptr) {
```
This prevents effects from being unloaded. This shouldn't be a debug assertion. In this case, call `unloadEffect()` then return. Since `unloadEffect()` is required when replacing an effect as well, you can start the function with just `unloadEffect()` outside of any `if` block, then return if the manifest is a nullptr.

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -299,15 +299,15 @@ EffectButtonParameterSlotPointer EffectSlot::getEffectButtonParameterSlot(unsign
+void EffectSlot::loadEffect(EffectManifestPointer pManifest, EffectInstantiatorPointer pInstantiator,
         const QSet<ChannelHandleAndGroup>& activeChannels) {
     // Remove an effect if already loaded in slot
     if (m_pEngineEffect != nullptr) {
```
remove this `if`. `unloadEffect()` is needed in every case.

kshitij98 replied:
I did not remove this to maintain a sanity check of calling unloadEffect only if the effectslot is already loaded. An assertion is present in removeFromEngine method for the same.

Be-ing replied:
Oh sorry, you are right.

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -150,72 +299,83 @@ EffectButtonParameterSlotPointer EffectSlot::getEffectButtonParameterSlot(unsign
-        // EffectPointer to the EffectSlot. Effects and EngineEffects default to
-        // disabled, so if this EffectSlot was enabled, enable the Effect and EngineEffect.
-        pEffect->setEnabled(m_pControlEnabled->toBool());
+    if (pManifest != nullptr || pInstantiator != nullptr) {
```
Add a comment: "In this case, only unload the old effect and return"

Be-ing replied:
This logic is backwards. Change to:
`if (pManifest == nullptr || pInstantiator == nullptr) {`

Be-ing replied:
also move `m_pManifest = pManifest` above this

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -379,7 +379,7 @@ void EffectChainSlot::enableForInputChannel(const ChannelHandleAndGroup& handle_
     for (int i = 0; i < m_effectSlots.size(); ++i) {
         auto& statesMap = (*pEffectStatesMapArray)[i];
-        if (m_effectSlots[i] != nullptr) {
+        if (m_effectSlots[i].isLoaded()) {
```
`m_effectSlots[i]->isLoaded()`

Be-ing committed:
use shared pointers for EngineEffectParameters

Be-ing committed:
decouple EffectProcessor construction from EngineEffect

This will allow for constructing EffectProcessors before constructing
EngineEffect.

Be-ing committed:
fix crash on shutdown when an LV2 effect was loaded

The crash was introduced in when removing EffectChainManager in
fded4f8d5b483681f19bbd0943baa806c8c8aad7

Be-ing committed:
make LV2EffectProcessor a subclass of EffectProcessorImpl

to reduce code duplication

Be-ing committed:
make LV2EffectManifest a subclass of EffectManifest

Be-ing committed:
activate and deactivate LV2 effect plugins appropriately

Be-ing committed:
remove unused function parameter

and use "override" where appropriate

Be-ing commented on src/effects/effectslot.h:
```diff
@@ -60,62 +74,79 @@ class EffectSlot : public QObject {
+    EffectParameter* getButtonParameterForSlot(unsigned int slotNumber);
+
+    EffectParameter* getParameterById(const QString& id) const;
+    EffectParameter* getButtonParameterById(const QString& id) const;
```
unused. Remove these functions.

Be-ing commented on src/effects/effectslot.h:
```diff
@@ -60,62 +74,79 @@ class EffectSlot : public QObject {
+    EffectManifestPointer m_pManifest;
+    EngineEffect* m_pEngineEffect;
+    QList<EffectParameter*> m_parameters;
+    QMap<QString, EffectParameter*> m_parametersById;
```
remove

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -150,72 +297,92 @@ EffectButtonParameterSlotPointer EffectSlot::getEffectButtonParameterSlot(unsign
+        EffectParameter* pParameter = new EffectParameter(
+                this, m_pEffectsManager, m_parameters.size(), pManifestParameter);
+        m_parameters.append(pParameter);
+        VERIFY_OR_DEBUG_ASSERT(!m_parametersById.contains(pParameter->id())) {
```
This is the only place `m_parametersById` is used. Remove it.

Be-ing commented on src/effects/effectslot.h:
```diff
@@ -60,62 +74,79 @@ class EffectSlot : public QObject {
+    double getMetaknobDefault();
+
   public slots:
     // Request that this EffectSlot load the given Effect
```
remove useless and out-of-place comment

Be-ing commented on src/effects/effectslot.h:
```diff
@@ -128,8 +159,6 @@ class EffectSlot : public QObject {
-    QList<EffectParameterSlotPointer> m_parameters;
-    QList<EffectButtonParameterSlotPointer> m_buttonParameters;
 
     SoftTakeover* m_pSoftTakeover;
```
rename `m_pMetaknobSoftTakeover`

Be-ing commented:
Replace the Qt `foreach` macro calls in EffectSlot with standard C++ `for` loop.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -192,6 +192,15 @@ void EffectChainSlot::loadEffect(const unsigned int iEffectSlotNumber,
+void EffectChainSlot::refreshAllEffects() {
+    for (EffectSlotPointer pSlot : m_effectSlots) {
+        if (pSlot->isLoaded()) {
+            pSlot->loadEffect(pSlot->getManifest(), pSlot->getInstantiator(), m_enabledInputChannels);
```
This makes EffectChainSlot and EffectSlot tightly coupled. Instead of using getter functions to access EffectSlot's private members, make a new EffectSlot::refresh function.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -192,6 +192,15 @@ void EffectChainSlot::loadEffect(const unsigned int iEffectSlotNumber,
 
+void EffectChainSlot::refreshAllEffects() {
+    for (EffectSlotPointer pSlot : m_effectSlots) {
+        if (pSlot->isLoaded()) {
```
EffectSlot::loadEffect checks this case already.

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -175,6 +175,10 @@ EffectManifestPointer EffectSlot::getManifest() const {
     return m_pManifest;
 }
 
+EffectInstantiatorPointer EffectSlot::getInstantiator() const {
```
remove

Be-ing commented on src/effects/effectsmanager.cpp:
```diff
@@ -401,24 +401,15 @@ void EffectsManager::setup() {
 void EffectsManager::refreshAllChainSlots() {
     for (auto& pChainSlot : m_standardEffectChainSlots) {
-        refreshChainSlot(pChainSlot);
+        pChainSlot->refreshAllEffects();
```
"refresh" doesn't clearly convey what this does. How about "reloadAllEffects"?

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -295,11 +294,12 @@ EffectButtonParameterSlotPointer EffectSlot::getEffectButtonParameterSlot(unsign
 void EffectSlot::loadEffect(EffectManifestPointer pManifest,
                             EffectInstantiatorPointer pInstantiator,
-                            const QSet<ChannelHandleAndGroup>& activeChannels) {
+                            QSet<ChannelHandleAndGroup>& activeChannels) {
```
revert

Be-ing commented on src/effects/effectslot.h:
```diff
@@ -140,6 +138,8 @@ class EffectSlot : public QObject {
     EffectsManager* m_pEffectsManager;
     EffectManifestPointer m_pManifest;
     EffectInstantiatorPointer m_pInstantiator;
+    QSet<ChannelHandleAndGroup> m_pActiveChannels;
```
Keep this state only in EffectChainSlot. Duplicating it here introduces the possibility that it will get out of sync with the state of EffectChainSlot in the future. Actually, this already happens when the active input channels are changed after an effect is loaded. You can verify that by:
1. Loading an effect
2. Enabling an input channel
3. Changing a checkbox in DlgPrefLV2
4. Pressing Apply. This crashes Mixxx.

Be-ing commented on src/effects/effectchainslot.cpp:
```diff
@@ -192,13 +192,10 @@ void EffectChainSlot::loadEffect(const unsigned int iEffectSlotNumber,
-        if (pSlot->isLoaded()) {
-            pSlot->loadEffect(pSlot->getManifest(), pSlot->getInstantiator(), m_enabledInputChannels);
-        }
+        pSlot->reload();
```
Keep passing `m_enabledInputChannels` here. It was only the EffectManifest and EffectInstantiator arguments that should have been removed. This keeps EffectSlot's state in EffectSlot only and EffectChainSlot in EffectChainSlot only.

Be-ing replied:
Requiring that argument should ensure that `EffectSlot::reload` is only ever called by EffectChainSlot. There will be a similar situation when implementing EffectPreset loading.

kshitij98 committed:
[WIP] Updated Effect tests

- SIGSEGV on loading an effect

Be-ing committed:
store EffectsBackends in a QHash indexed by EffectBackendType

... in EffectsManager. This will allow for instantiating effects
properly with just an EffectManifest, by first fetching the appropriate
EffectBackend based on the EffectBackendType stored on the
EffectManifest. Currently EffectsManager just iterates through the
manifests and asks the backend if it can instantiate an effect ID. This
is wrong because the behavior is undefined in the event that two
EffectManifests with different backends happen to share the same effect
ID.

Be-ing committed:
pass EffectProcessor instead of EffectInstantiator to EngineEffect

and move EffectInstantiator where it belongs, as a private
implementation detail of BuiltInBackend. Move the implementation of
EffectsBackend into BuiltInBackend and leave EffectsBackend as an
abstract base class.

Be-ing committed:
use EffectManifests instead of QString IDs to identify effects

Effect ID strings are only guaranteed to be unique for one
EffectsBackend. Using only the ID strings to identify effects creates
the possibility of a name collision between different effect types from
different EffectsBackends. EffectManifests contain both the ID string
and the EffectBackendType so they uniquely identify an effect type.

This refactoring forced a big cleanup of DlgPrefEq. Lots of duplicate
code was removed from that class, but there is still tight coupling
between DlgPrefEq and EffectChainSlot + EffectSlot for the master
equalizer.

(cherry picked from commit 59bdce0bc2a792d375a428baf4e628f640ef523b)

kshitij98 committed:
Added Qt 5.5.1 support

kshitij98 committed:
Removed redundancy from EffectParameter and EffectParameterSlot

- Removed wrapper functions from EffectParameter
- Removed clamping from EffectParameter and added assertions directly to EffectManifestParameter
- Added setRange function instead of setMinimum, setDefault & setMaximum so that order of setting does not matter and invariancy in ensured

Be-ing commented on src/effects/effectmanifestparameter.h:
```diff
@@ -202,21 +204,22 @@ class EffectManifestParameter {
     }
-    virtual void setMaximum(const double& maximum) {
+
+    virtual void setRange(const double& minimum, const double& defaultValue, const double& maximum) {
```
Good idea to combine these into one function to assure all values are valid. :+1:

Be-ing commented on src/effects/effectmanifestparameter.h:
```diff
@@ -202,21 +204,22 @@ class EffectManifestParameter {
-        m_minimum = minimum;
-    }
 
     virtual const double& getMaximum() const {
```
Can you remove the `virtual` keyword from these functions? There aren't any subclasses of EffectManifestParameter and I doubt there ever will be.

Be-ing commented:
Now that you've removed a lot of clutter from EffectParameter, I think moving what is left of that class into EffectParameterSlot should be pretty straightforward (easier than it was to combe EffectChain/EffectChainSlot or Effect/EffectSlot).

Be-ing commented on src/effects/effectbuttonparameterslot.cpp:
```diff
@@ -23,7 +23,7 @@ EffectButtonParameterSlot::EffectButtonParameterSlot(const QString& group,
             ConfigKey(m_group, itemPrefix + QString("_type")));
 
     connect(m_pControlValue, SIGNAL(valueChanged(double)),
```
move to EffectParameterSlotBase

kshitij98 replied:
We cannot shift `m_pControlValue` to EffectParameterSlotBase.

Be-ing commented on src/effects/effectbuttonparameterslot.cpp:
```diff
@@ -38,82 +38,85 @@ EffectButtonParameterSlot::~EffectButtonParameterSlot() {
+
+        // Set the number of states
+        int numStates = math_max(m_pManifestParameter->getSteps().size(), 2);
+        m_pControlValue->setStates(numStates);
```
I think this is the only part of this function that is not shared with EffectParameterSlot. Code duplication can be reduced by moving all the common code to EffectParameterSlotBase. In this EffectButtonParameterSlot::loadEffect function, call EffectParameterSlotBase::loadEffect, then put the code specific to this subclass.

kshitij98 replied:
I tried shifting loadManifestParameter to EffectParameterSlotBase, But only a few lines could be shifted, so I have left this function in the subclasses only.

kshitij98 commented on src/effects/effectparameterslot.cpp:
```diff
@@ -56,62 +56,70 @@ EffectParameterSlot::~EffectParameterSlot() {
 
-void EffectParameterSlot::loadEffect(EffectSlot* pEffectSlot) {
-    // qDebug() << debugString() << "loadEffect" << (pEffectSlot ? pEffectSlot->getManifest().name() : "(null)");
+void EffectParameterSlot::updateEngineState() {
```
No, we cannot move this to EffectParameterSlotBase because updateEngineState function uses `m_pControlValue`. We cannot declare `m_pControlValue` in EffectParameterSlotBase because it's datatype depends on whether it is knob parameter or button parameter. We can make a `EffectParameterSlotBase::updateEngineState(double value)` method and declare wrapper functions in EffectParameterSlot and EffectButtonParameterSlot.

Be-ing replied:
Are you sure? `m_pControlValue` is a subclass of `ControlObject` in both `EffectParameterSlotBase` subclasses, so I think you can declare it with the `ControlObject` type in `EffectParameterSlotBase` and initialize it as nullptr in the `EffectParameterSlotBase` member initialization list. Then the `EffectParameterSlotBase` subclass constructors could initialize `m_pControlValue` with the appropriate `ControlObject` subclass and make the Qt signal-slot connection.

Be-ing commented:
There is a merge conflict with master in src/preferences/dialog/dlgprefeq.cpp.

kshitij98 commented:
> There is a merge conflict with master in src/preferences/dialog/dlgprefeq.cpp.

It is conflicting with unique ID [commit](https://github.com/mixxxdj/mixxx/pull/1705/commits/3f61af9eec6b5cd6d5e3945c1d3a0fb654c1d0d3). Can you resolve the same?

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -313,13 +293,11 @@ void EffectSlot::loadEffect(const EffectManifestPointer pManifest,
     for (const auto& pManifestParameter: m_pManifest->parameters()) {
         if (isKnobParameter(pManifestParameter)) {
-            m_knobParameterSlots[iParameterSlot]->loadManifestParameter(
+            m_knobParameterSlots[iParameterSlot++]->loadManifestParameter(
```
Please do not use this antifeature of C++ that lets you assign to a variable and reference it on the same line. It makes the code unnecesarily difficult to think about. Put `++iParameterSlot;` back on its own line.

kshitij98 replied:
Updated.

Be-ing commented on src/effects/effectbuttonparameterslot.cpp:
```diff
@@ -17,14 +17,16 @@ EffectButtonParameterSlot::EffectButtonParameterSlot(EffectsManager* pEffectsMan
-
     connect(m_pControlValue, SIGNAL(valueChanged(double)),
             this, SLOT(updateEngineState()));
+    ControlPushButton* pControlValue = static_cast<ControlPushButton*>(m_pControlValue);
```
Huh, I didn't expect this to work because I did not realize `static_cast` implicitly calls the constructor. Please use `m_pControlValue = new ControlPushButton` instead for clarity.

Be-ing commented on src/effects/effectbuttonparameterslot.h:
```diff
@@ -29,23 +29,14 @@ class EffectButtonParameterSlot : public EffectParameterSlotBase {
     }
 
-    // Control exposed to the rest of Mixxx
-    ControlPushButton* m_pControlValue;
```
Correct me if I'm wrong, but I think you still need this to access the methods of ControlPushButton in EffectButtonParameterSlot that are not common to ControlObject.

kshitij98 replied:
This is required to move updateEngineState to EffectParameterSlotBase. We can use ControlPushButton specific methods by typecasting.

Be-ing commented on src/effects/effectknobparameterslot.cpp:
```diff
@@ -97,7 +83,8 @@ void EffectKnobParameterSlot::loadManifestParameter(unsigned int iParameterNumbe
 
         EffectManifestParameter::ControlHint type = m_pManifestParameter->controlHint();
-        m_pControlValue->setBehaviour(type, dMinimum, dMaximum);
+        ControlEffectKnob* pControlValue = static_cast<ControlEffectKnob*>(m_pControlValue);
```
This casting shouldn't be needed if you have a ControlEffectKnob member variable.

kshitij98 replied:
Then we won't be able to use the ControlObject objects in EffectParameterSlotBase. Or wait, not if we can have virtual member variables also.

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -313,18 +313,6 @@ void EffectSlot::loadEffect(const EffectManifestPointer pManifest,
 
     m_pControlLoaded->forceSet(1.0);
 
-    unsigned int iNumKnobParameters = numKnobParameters();
```
I was wondering about removing this code. Thanks for going ahead and doing it. The number of parameter slots should be fixed, otherwise skins and controller mappings don't know how many they may possibly need to support.

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -144,10 +144,10 @@ void EffectSlot::updateEngineState() {
     }
     sendParameterUpdate();
-    for (auto const& pParameterSlot : m_parameterSlots) {
+    for (auto const& pParameterSlot : m_knobParameterSlots) {
```
Thanks, this makes the code easier to understand. :+1:

Be-ing commented on build/depends.py:
```diff
@@ -758,8 +758,6 @@ def sources(self, build):
                    "effects/effectmanifest.cpp",
                    "effects/effectmanifestparameter.cpp",
 
-                   "effects/effectparameter.cpp",
```
:tada:

Be-ing commented on build/depends.py:
```diff
@@ -762,7 +762,7 @@ def sources(self, build):
                    "effects/effectslot.cpp",
                    "effects/effectparameterslotbase.cpp",
-                   "effects/effectparameterslot.cpp",
+                   "effects/effectknobparameterslot.cpp",
```
Thanks, this makes the code easier to understand :+1:

Be-ing commented on src/effects/effectparameterslotbase.cpp:
```diff
@@ -79,4 +79,13 @@ void EffectParameterSlotBase::clear() {
+    if (m_pManifestParameter && m_pManifestParameter->showInParameterSlot()) {
+        m_pControlLoaded->forceSet(1.0);
+    } else {
+        m_pControlLoaded->forceSet(0.0);
```
This alone breaks controller mappings. Hidden parameters do need to be unloaded from the parameter slot. Merely changing `m_pControlLoaded` does hide parameters from skins, but the ControlObjects are still used by hidden parameters. For example, if the first parameter is hidden, a controller knob mapped to `[EffectRack1_EffectUnitX_EffectY], parameter1` still controls the hidden parameter instead of manipulating the first parameter shown on screen.

Be-ing commented on src/effects/effectparameterslotbase.cpp:
```diff
@@ -79,4 +79,13 @@ void EffectParameterSlotBase::clear() {
+    } else {
+        m_pControlLoaded->forceSet(0.0);
+    }
+    emit(updated());
```
This signal name is unhelpful. How about renaming it to `parameterLoaded`?

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -253,13 +259,15 @@ void EffectSlot::setEnabled(bool enabled) {
-EffectKnobParameterSlotPointer EffectSlot::getEffectKnobParameterSlot(unsigned int slotNumber) {
+EffectKnobParameterSlotPointer EffectSlot::getEffectKnobParameterSlot(unsigned int position) {
     //qDebug() << debugString() << "getEffectKnobParameterSlot" << slotNumber;
+    int slotNumber = m_knobParameterPositionToSlot.value(position, -1);
```
Revert this. When hidden parameters are not loaded in parameter slots, this would require other classes to know which parameters are hidden to get the parameter slot they are searching for. 

Eventually this function should be removed entirely because other classes should not have direct access to parameter slots. All access to the effects system from outside the effects system should be mediated by EffectsManager to decouple the GUI from implementation details. Please add a comment noting this.

The same applies for getEffectButtonParameterSlot.

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -304,6 +312,20 @@ void EffectSlot::loadEffect(const EffectManifestPointer pManifest,
         ++iNumParameter;
     }
 
+    for (unsigned int i=0 ; i<m_knobParameterSlots.size() ; ++i) {
```
Remove this new code. It will not be needed when hidden parameters are not loaded to slots.

Be-ing commented on src/effects/effectslot.h:
```diff
@@ -133,6 +133,8 @@ class EffectSlot : public QObject {
     QList<EffectKnobParameterSlotPointer> m_knobParameterSlots;
     QList<EffectButtonParameterSlotPointer> m_buttonParameterSlots;
+    QList<int> m_knobParameterPositionToSlot;
+    QList<int> m_buttonParameterPositionToSlot;
```
These names do not make it clear what the "position" referred to is. Also, I think it is backwards; a map of slot indices to manifest indices would be more helpful for passing that information to the parameter slot.

Be-ing reviewed (changes requested):
I like that you tried to avoid reloading the whole effect to implement parameter reordering, but this implementation does not work for controllers. Hidden parameters should not be loaded in any slot. Parameter slots need to be passed an int representing the loaded parameter's position in the EffectManifest so EffectParameterSlotBase::updateEngineState can send a message to the engine for the correct parameter (this will keep hiding and rearrangement entirely transparent to the engine).

Please consolidate the common code in the loadManifestParameter functions into EffectParameterSlotBase. It will be a pain to keep updating the same function in multiple classes and a bug could easily sneak in by forgetting some detail in one class.

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -178,16 +179,20 @@ EffectManifestPointer EffectSlot::getManifest() const {
-void EffectSlot::reload(const QSet<ChannelHandleAndGroup>& activeInputChannels) {
-    loadEffect(m_pManifest,
-            m_pEffectsManager->createProcessor(m_pManifest),
-            activeInputChannels);
```
perhaps instead of calling EffectSlot::loadEffect here, you could move the part of EffectSlot::loadEffect that loads the parameter slots to a new private function and call it both here and from loadEffect

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -178,16 +179,20 @@ EffectManifestPointer EffectSlot::getManifest() const {
+    }
+    for (auto const& pParameterSlot : m_buttonParameterSlots) {
+        pParameterSlot->reload();
+    }
```
There are several places in EffectSlot where code is duplicated for knob and button parameters. It would help to store a QList<EffectParameterSlotBasePointer> and iterate over that instead.

kshitij98 committed:
Added EffectParameter and EffectParameterSlot class comments

Be-ing commented on src/effects/effectparameter.cpp:
```diff
@@ -2,38 +2,20 @@
-EffectParameter::EffectParameter(Effect* pEffect, EffectsManager* pEffectsManager,
-                                 int iParameterNumber,
-                                 EffectManifestParameterPointer pParameter)
+EffectParameter::EffectParameter(EffectSlot* pEffectSlot, EffectsManager* pEffectsManager,
```
The only argument this needs in the EffectManifestParameterPointer.

Be-ing commented on src/effects/effectparameter.h:
```diff
@@ -5,59 +5,27 @@
-// always within acceptable ranges. This class is NOT thread-safe and must only
-// be used from the main thread.
+// An EffectParameter is a wrapper around EffectManifestParameter that tracks a
+// mutable value state and communicates that state to the engine. This class is
```
After the first sentence, add "Separating this from the parameterX ControlObjects in EffectParameterSlot allows for decoupling the state of the parameters from the ControlObject states, which is required for parameter hiding and rearrangement."

Be-ing replied:
ping

Be-ing commented on src/effects/effectparameter.h:
```diff
@@ -5,59 +5,27 @@
-    // Parameter Information
-    ///////////////////////////////////////////////////////////////////////////
-
     EffectManifestParameterPointer manifest() const;
```
rename to `getParameterManifest` to avoid potential confusion with EffectManifest

Be-ing commented on src/effects/effectparameter.h:
```diff
@@ -5,59 +5,27 @@
-
-    EffectManifestParameter::ControlHint getControlHint() const;
-
     void updateEngineState();
```
Should this be public? I'm not sure.

Be-ing commented on src/effects/effectparameter.h:
```diff
@@ -71,18 +39,12 @@ class EffectParameter : public QObject {
 
-    Effect* m_pEffect;
+    EffectSlot* m_pEffectSlot;
     EffectsManager* m_pEffectsManager;
```
remove

Be-ing commented on src/effects/effectparameter.h:
```diff
@@ -71,18 +39,12 @@ class EffectParameter : public QObject {
-    bool clampRanges();
 
-    Effect* m_pEffect;
+    EffectSlot* m_pEffectSlot;
```
remove

Be-ing commented on src/effects/effectparameter.h:
```diff
@@ -71,18 +39,12 @@ class EffectParameter : public QObject {
-    Effect* m_pEffect;
+    EffectSlot* m_pEffectSlot;
     EffectsManager* m_pEffectsManager;
     int m_iParameterNumber;
```
remove

Be-ing commented on src/effects/effectparameter.h:
```diff
@@ -5,59 +5,27 @@
-
     void updateEngineState();
 
   signals:
```
remove

Be-ing commented on src/effects/effectparameter.h:
```diff
@@ -5,59 +5,27 @@
+// An EffectParameter is a wrapper around EffectManifestParameter that tracks a
+// mutable value state and communicates that state to the engine. This class is
+// NOT thread-safe and must only be used from the main thread.
 class EffectParameter : public QObject {
```
do not subclass QObject

Be-ing commented on src/effects/effectparameter.h:
```diff
@@ -5,59 +5,27 @@
 #include <QVariant>
 
 #include "effects/effectmanifestparameter.h"
+#include "effects/effectslot.h"
```
remove

Be-ing commented on src/effects/effectparameter.h:
```diff
@@ -5,59 +5,27 @@
 #include <QVariant>
```
remove

Be-ing commented on src/effects/effectparameter.h:
```diff
@@ -5,59 +5,27 @@
+#include "effects/effectslot.h"
 #include "util/class.h"
 
 class Effect;
```
remove

Be-ing commented on src/effects/effectparameter.h:
```diff
@@ -5,59 +5,27 @@
 #include "util/class.h"
 
 class Effect;
 class EffectsManager;
```
remove

kshitij98 committed:
Renamed EffectParameterSlotPointer to EffectKnobParameterSlotPointer

kshitij98 committed:
Renamed m_parameterSlots -> m_knobParameterSlots

kshitij98 committed:
Renamed EffectParameterSlot -> EffectKnobParameterSlot

kshitij98 committed:
Renamed m_buttonParameters -> m_buttonParameterSlots

kshitij98 committed:
Effect Parameter Hiding/Rearrangement

Be-ing commented on src/preferences/dialog/dlgpreflv2.cpp:
```diff
@@ -117,11 +117,15 @@ void DlgPrefLV2::slotApply() {
-        EffectManifestParameterPointer pParameter = pCurrentEffectManifest->parameter(i);
-        pParameter->setShowInParameterSlot(m_pluginParameters[i]->isChecked());
+        if (m_pluginParameters[i]->isChecked()) {
+            m_pEffectsManager->setEffectParameterPosition(pCurrentEffectManifest, i, position);
```
Unfortunately this only changes effects that are currently loaded. If the LV2 plugin is loaded again later, all its parameters are loaded. As a quick solution we may consider going back to the ugly hack of setting a flag on the EffectManifestParameter and checking that flag in EffectSlot::loadEffect. In the long term I think we should implement [custom default EffectPresets for each effect type](https://bugs.launchpad.net/mixxx/+bug/1740504), but of course that requires having EffectPresets implemented first.

kshitij98 replied:
Yes, I've added a TODO to remove these functions once the GUI is implemented for this. Once we have a GUI for hiding effect parameters, We'll be able to hide parameters for particular instances of EffectSlots.

Be-ing reviewed (changes requested):
This is a good start for parameter hiding and rearrangement! There are a few issues to fix though.

kshitij98 commented on src/effects/effectknobparameterslot.cpp:
```diff
@@ -56,51 +56,50 @@ EffectKnobParameterSlot::~EffectKnobParameterSlot() {
+        m_pControlLinkType->set(
+            static_cast<double>(m_pManifestParameter->defaultLinkType()));
+        m_pControlLinkInverse->set(
+            static_cast<double>(m_pManifestParameter->defaultLinkInversion()));
```
I think we should aim at shifting minimum, maximum, default, current value, link type, link inversion etc, in EffectParameter because that will allow us to set custom defaults, ranges, ... for that instance.

kshitij98 replied:
Also, I can't foresee the purpose of dMinimumLimit and dMaximumLimit. Should we not shift the warning/assertion to their respective classes than in Effect*ParameterSlot::loadParameter?

kshitij98 committed:
Reduced Effect Parameter Code Duplicacy

- Added ParameterType enum and merged Knob/Button methods

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -175,40 +159,23 @@ EffectManifestPointer EffectSlot::getManifest() const {
 }
 
-unsigned int EffectSlot::numButtonParameters() const {
+unsigned int EffectSlot::numParameters(EffectParameterType parameterType) const {
```
This function is unused. Other classes should not have access to this implementation detail anyway. Please remove this function.

kshitij98 replied:
It is being used in DlgPrefEq class. I think we can implement a `nextParameter(EffectParameterType parameterType, int index)` method in EffectSlot and remove this function. What do you think?

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -398,22 +342,22 @@ void EffectSlot::unloadEffect() {
-    m_knobParameterSlotPositionToManifestIndex.removeAll(parameterId);
-    m_buttonParameterSlotPositionToManifestIndex.removeAll(parameterId);
+    m_parameterSlotPositionToManifestIndex[EffectParameterType::Knob].removeAll(parameterId);
+    m_parameterSlotPositionToManifestIndex[EffectParameterType::Button].removeAll(parameterId);
```
Iterate over the values of m_parameterSlotPositionToManifestIndex like
```
for (auto& map : m_parameterSlotPositionToManifestIndex) {
    map.removeAll(index);
}
```

Be-ing reviewed (changes requested):
Thanks for continuing to work on this after GSOC has officially ended. I like the introduction of a new enum class for parameter type. I think we could use it clarify the conflation of EffectManifestParameter::ControlHint with the parameter type.

kshitij98 committed:
Renamed ControlHint->ValueScaler and EffectParameterType->ParameterType

kshitij98 commented on src/effects/effectslot.cpp:
```diff
@@ -297,18 +256,18 @@ void EffectSlot::loadEffect(const EffectManifestPointer pManifest,
         } else {
-            m_buttonParameterSlotPositionToManifestIndex.push_back(index);
+            m_parameterSlotPositionToManifestIndex[EffectParameterType::Button].push_back(index);
         }
```
For now, I've used ControlHint::TOGGLE_STEPPING to set the parameter type in effect manifest. I think we can remove it and explicitly set the parameter type in each builtin effect by calling `manifest->setParameterType(...)`. What do you think?

Be-ing replied:
I agree, let's have each parameter's type explicitly specified.

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -29,18 +29,6 @@ EffectSlot::EffectSlot(const QString& group,
     m_pControlLoaded = new ControlObject(ConfigKey(m_group, "loaded"));
     m_pControlLoaded->setReadOnly();
 
-    m_pControlNumParameters = new ControlObject(ConfigKey(m_group, "num_parameters"));
```
ping

kshitij98 replied:
Updated

kshitij98 committed:
Added ControlNumParameters

Be-ing commented on src/effects/effectknobparameterslot.cpp:
```diff
@@ -44,67 +49,73 @@ EffectParameterSlot::EffectParameterSlot(const QString& group, const unsigned in
+            static_cast<double>(m_pManifestParameter->defaultLinkInversion()));
+
+        connect(m_pEffectParameter, SIGNAL(valueChanged(double)),
+                this, SLOT(slotParameterValueChanged(double)));
```
I'm wondering if this should be removed. This creates an overcomplicated code path when a parameter ControlObject is updated. First `EffectKnobParameterSlot::slotValueChanged` calls `m_pEffectParameter->setValue`, which then emits the `valueChanged` signal, then the parameter slot updates the ControlObject... but the ControlObject is already up-to-date because that's what triggered `slotValueChanged` to start with.

I don't think any class except the parameter slots should be setting the value of EffectParameters directly. DlgPrefEq does this in a few places and I think it would be better to have DlgPrefEq working with the parameter slots instead. This would also allow for removing `EffectSlot::getParameterForSlot`. Also, removing the `valueChanged` signal from EffectParameter would allow EffectParameter to not inherit from QObject.

kshitij98 replied:
Yes, I've removed directly setting the EffectParameter and now the corresponding ParameterSlot is making those updates. This allowed us to remove a few redundant methods.
https://github.com/mixxxdj/mixxx/pull/1705/commits/d7d565791124e03a94e64478a21e34a27018d092

Be-ing replied:
Nice work :+1:

Be-ing commented on src/effects/effectknobparameterslot.h:
```diff
@@ -52,7 +54,7 @@ class EffectParameterSlot : public EffectParameterSlotBase {
+        return QString("EffectKnobParameterSlot(%1,%2)").arg(m_group).arg(m_iParameterSlotNumber);
     }
 
     SoftTakeover* m_pSoftTakeover;
```
rename to `m_pMetaknobSoftTakeover` for clarity

Be-ing commented on src/effects/effectknobparameterslot.cpp:
```diff
@@ -171,7 +182,7 @@ void EffectParameterSlot::onEffectMetaParameterChanged(double parameter, bool fo
                     return;
                 }
                 {
```
There is no need to have this in its own block. Delete the unnecessary block and unindent the following code.

kshitij98 replied:
It is their in it's own block because of the local variable declaration of `double neutral` inside the switch case. I've kept it as it is for now. Otherwise we can declare `neutral` outside of the switch case statement and remove the block.

Be-ing replied:
Why does that variable need its own block scope though? It doesn't conflict with anything else in the function's scope, does it?

Be-ing commented on src/effects/effectbuttonparameterslot.cpp:
```diff
@@ -113,33 +119,39 @@ void EffectButtonParameterSlot::slotValueChanged(double v) {
+    // }
+}
+
+void EffectButtonParameterSlot::syncSofttakeover() {
```
Instead of having somewhat confusing blank functions in this class for `syncSofttakeover` and `onEffectMetaParameterChanged`, I think it would be better to have the default implementations of these functions in EffectParameterSlotBase be empty functions and override the default implementation in EffectKnobParameterSlot. That way, EffectButtonParameterSlot wouldn't need any code relating to these functions.

Be-ing commented on src/effects/effectbuttonparameterslot.h:
```diff
@@ -5,14 +5,16 @@
 #include <QVariant>
 #include <QString>
 
+// #include "effects/defs.h"
```
remove commented code

Be-ing reviewed (changes requested):
I did a thorough review of all of effectslot.cpp, so some of these comments are not directly related to your most recent changes. After taking care of these comments, I think we'll be ready to move on to reimplementing XML preset saving/loading.

Be-ing commented:
Hi @kshitij98, do you think you'll have time to continue working on this soon? A few merge conflicts have developed with master.

kshitij98 committed:
Moved to initialiser list
Removed virtual

kshitij98 commented:
> Hi @kshitij98, do you think you'll have time to continue working on this soon? A few merge conflicts have developed with master.

Hi @Be-ing, I just picked up from where I left. I'll resolve the emerged conflicts very soon.

kshitij98 commented on src/effects/effectslot.cpp:
```diff
@@ -68,154 +80,301 @@ EffectSlot::EffectSlot(const QString& group,
 
-void EffectSlot::slotEffectEnabledChanged(bool enabled) {
-    m_pControlEnabled->set(enabled);
+EngineEffect* EffectSlot::getEngineEffect() {
```
This function is being used in EffectParameter::updateEngineState, I've not removed it yet.

Be-ing replied:
I looked through EffectParameter and this is the only reason it needs a pointer to EffectSlot. I think it would be better to pass the EffectParameter constructor a pointer to the EngineEffect than EffectSlot, then this function could be removed.

kshitij98 commented on src/effects/effectslot.cpp:
```diff
@@ -68,154 +80,301 @@ EffectSlot::EffectSlot(const QString& group,
+        return;
+    }
+
+    for (auto& parameterMapping : m_parameterSlotPositionToManifestIndex) {
```
mapForParameterType sounds good!

kshitij98 committed:
Renamed m_parameterSlotPositionToManifestIndex -> m_mapForParameterType

kshitij98 committed:
Removed spacing between "> ("

kshitij98 commented on src/effects/effectslot.cpp:
```diff
@@ -68,154 +80,301 @@ EffectSlot::EffectSlot(const QString& group,
+
+            auto pParameter = m_parameters.value(iParameterPosition, nullptr);
+
+            VERIFY_OR_DEBUG_ASSERT(pParameter != nullptr) {
```
I think this is because of the previously unclear naming. It should have been iParameterSlot. I have renamed it to parameterSlotIndex now.

Not incrementing the same here enables us to skip invalid parameter manifests without wasting a parameter slot.

Be-ing replied:
Okay, that makes sense. It would be good to add a comment here explaining that.

kshitij98 committed:
Renamed iParameter -> parameterSlotIndex
Renamed iParameterPosition -> manifestIndex

kshitij98 committed:
Cleared Parameter Mappings

kshitij98 commented on src/effects/effectslot.cpp:
```diff
@@ -68,154 +80,301 @@ EffectSlot::EffectSlot(const QString& group,
+                continue;
+            }
+
+            if (iParameter < numParameterSlots) {
```
Earlier I was breaking out of the loop whenever an invalid parameter slot occurred. I have added the intended functionality in the latest [commit](https://github.com/mixxxdj/mixxx/pull/1705/commits/6aafc7159a08c3098d2fec6548bf2cccf4c2a8cc).

kshitij98 committed:
Added TODO: reimplement slotEffectSelector

kshitij98 committed:
Renamed m_pSoftTakeover -> m_pMetaknobSoftTakeover
Removed unnecessary range check in parameters

kshitij98 committed:
Fixed inheritence in parameter slots
Removed syncSofttakeover
Removed onEffectMetaParameterChanged

Be-ing commented:
Please write more descriptive commit messages. Even something as simple as "rename some variables in EffectSlot" is better.

kshitij98 commented:
> Please write more descriptive commit messages. Even something as simple as "rename some variables in EffectSlot" is better.

I have edited the previous commit messages

kshitij98 committed:
Merge branch 'master' into effects_refactoring

Be-ing committed:
function pointers for signal-slot connections in effects system

Be-ing commented:
https://github.com/kshitij98/mixxx/pull/10

daschuer reviewed (changes requested):
Unfortunately this PR is still blocked by the open discussion about the future CO namespace. We are talking here about 8000 alias COs.

I understand the reasons for the renaming, but we should pretty sure that this reason outweighs all the related work and that the new namespace helps us for future changes. Later renaming because of new features will be harder due to the created naming legacy.  

I think both is doubtful, and the old name space is well considered and not that bad. I also do not like the idea of cycling the names, this will create even more confusion. Unit and Rack means here totally different things. A perfect recipe for confusion.

If it turns out I am the only one with this opinion we can do the change, otherwise lets postpone the renaming and revert this PR to the existing nomenclature.

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -316,33 +316,36 @@ void EffectSlot::loadParameters() {
-        unsigned int iParameter = 0;
+        unsigned int parameterSlotIndex = 0;
         unsigned int numParameterSlots = m_iNumParameterSlots[parameterType];
         for (int i=0 ; i<m_mapForParameterType[parameterType].size() ; ++i) {
```
Add a comment above this line: `Load EffectParameters into the slot indicated by m_mapForParameterType`.

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -316,33 +316,36 @@ void EffectSlot::loadParameters() {
         }
-        while (iParameter < numParameterSlots) {
-            auto pParameterSlot = getEffectParameterSlot(parameterType, iParameter);
+        while (parameterSlotIndex < numParameterSlots) {
```
Add a comment above this line: `Clear any EffectParameterSlots still have a loaded parameter from before and the loop above did not load a new parameter into them`

daschuer commented:
So the current state here is that the Rack classes where removed, but the CO namespace is still unchanged? If we stick with this for a first step, it will work for me.

Be-ing commented:
Yes, that is the current state of the code.

kshitij98 committed:
Removed redundant block from switch case

kshitij98 committed:
Added comment in load Parameter

kshitij98 committed:
Removed EffectSlot::getEngineEffect()
Added comments

kshitij98 commented on src/effects/effectslot.cpp:
```diff
@@ -316,33 +316,36 @@ void EffectSlot::loadParameters() {
 
-            if (iParameter < numParameterSlots) {
-                auto pParameterSlot = getEffectParameterSlot(parameterType, iParameter);
+            while (parameterSlotIndex < numParameterSlots) {
```
I changed it to a `while` loop to skip all the invalid parameter slots without skipping to the next parameter. `parameterSlotIndex` shouldn't be greater than `numParameterSlots`, but it is still possible if we have some invalid parameter slots in between.

Be-ing replied:
What do you mean by an invalid parameter slot? If we're assuming that should never happen, I think it should be documented with `VERIFY_OR_DEBUG_ASSERT`.

kshitij98 replied:
Ohh yes, I have a pushed the [latest commit](https://github.com/mixxxdj/mixxx/pull/1705/commits/c672bbdcad87e615e5340a8ce7bb6f5d3151b48d) to resolve the issue.

kshitij98 committed:
Merge pull request #10 from Be-ing/effects_refactoring

function pointers for signal-slot connections in effects system

kshitij98 committed:
Added assertion parameterSlotIndex < numParameterSlots

kshitij98 committed:
Added comment in EffectParameter

Be-ing commented on src/effects/effectslot.cpp:
```diff
@@ -68,154 +80,301 @@ EffectSlot::EffectSlot(const QString& group,
+        parameterMapping.removeAll(parameterId);
+    }
+
+    auto pParameter = m_parameters.at(parameterId);
```
I'm not sure we can depend on parameters being listed with all knobs then all buttons. For Mixxx's built in effects, we could easily document it should be that way. But I don't think we can rely on that for external effects plugins.

Be-ing replied:
I think we should rework this function to be more robust or somehow enforce the constraints it on EffectManifests that it assumes.

kshitij98 replied:
Maybe we can reorder the effect parameters according to the parameter type after loading an `EffectManifest` and overwrite the same?

Be-ing replied:
I'm leaning towards refactoring this function to take the EffectManifestParameter::ParameterType as an argument. We already store a separate map for each parameter type.

Currently this function is only used by DlgPrefLV2 (indirectly via EffectsManager and EffectChainSlot) for hiding LV2 parameters. DlgPrefLV2 does not distinguish between types of parameters and making it do that would add more complexity to the code. I think at this point we should go ahead and remove DlgPrefLV2. I think it would be more worthwhile to get started working towards saving custom per-effect default snapshots with XML than continue maintaining the hacky way hiding LV2 parameters was implemented before.

kshitij98 committed:
Shifted slotValueChanged to EffectParameterSlotBase

kshitij98 committed:
Simplified parameter value updates
- Removed EffectSlot::getParameterForSlot
- Removed EffectParameter::valueChanged signal

kshitij98 committed:
Merge remote-tracking branch 'upstream/master' into effects_refactoring

kshitij98 committed:
Added Effect Preset classes
- Added EffectChainPreset class
- Added EffectPreset class
- Added EffectsManager::loadEffectChainPresets

Be-ing commented on src/effects/effectpreset.cpp:
```diff
@@ -0,0 +1,32 @@
+	}
+
+    m_id = XmlParse::selectNodeQString(element, EffectXml::EffectId);
+	m_version = XmlParse::selectNodeQString(element, EffectXml::EffectVersion);
```
tab -> spaces

Be-ing commented on src/effects/effectchainpreset.h:
```diff
@@ -0,0 +1,24 @@
+	}
+
+  private:
+  	QString m_name;
```
Add
```
double m_dSuperParameter
EffectChainMixMode m_mixMode
```

Be-ing commented on src/effects/effectchainpreset.h:
```diff
@@ -0,0 +1,24 @@
+
+  private:
+  	QString m_name;
+
```
extra newline

Be-ing commented on src/effects/effectxmlelements.h:
```diff
@@ -6,9 +6,6 @@ const int kXmlSchemaVersion = 0;
 const QString Root("MixxxEffects");
 const QString SchemaVersion("SchemaVersion");
 
-const QString Rack("Rack");
```
This is still required for backward compatibility loading effects.xml files from older versions of Mixxx.

Be-ing reviewed (changes requested):
This is off to a good start!

Be-ing commented on src/effects/effectsmanager.cpp:
```diff
@@ -498,3 +496,32 @@ void EffectsManager::collectGarbage(const EffectsRequest* pRequest) {
+    file.close();
+
+    QDomElement root = doc.documentElement();
+    QDomElement rackElement = XmlParse::selectElement(root, EffectXml::Rack);
```
Add a comment explaining that this element is superfluous but it is kept for backwards compatibility.

Be-ing commented:
continued here: #2618
